wp/lib/bap_wp/src/z3_utils.ml

```diff
@@ -110,3 +110,160 @@ let mk_smtlib2_single (env : Env.t) (smt_post : string) : Constr.t =
+    ] *)
+
+let asserts_of_model (model_string : string) (sym_names : string list) : Sexp.t list =
+  let process_decl l = match l with
+    | Sexp.List [Sexp.Atom "define-fun" ; Sexp.Atom varname; args ; _ ; model_val ] ->
+      (* If variable is not in sym_names, z3 will crash on processing it.
+         Z3 can fill in gaps in the model via it's own smt search.
+         The main cost is slowdown. *)
+      if (List.mem sym_names varname ~equal:String.equal)
+      then
+        let model_val = match args with
+          | Sexp.List [] ->  (* Constant case *)
+            model_val
+          | Sexp.List _  ->  (* Function model *)
+            Sexp.List [Sexp.Atom "lambda" ; args; model_val] (* Sexp.Atom varname *)
+          | Sexp.Atom _ -> failwith "Unexpected atom"
+        in
+        Sexp.List [Sexp.Atom "assert" ;
+                   Sexp.List [Sexp.Atom "=" ; Sexp.Atom varname ; model_val]]
+      else
+        begin
+          Printf.printf "Warning: %s not instantiated in Z3 query\n" varname;
+          Sexp.List [Sexp.Atom "assert" ; Sexp.Atom "true"]
+        end
+    | _ -> failwith "Error: Unexpected form in external smt model"
+  in
+  let model_sexp = Sexp.of_string model_string in
+  match model_sexp with
+  | Sexp.List (Sexp.Atom "model" :: t) ->
+    List.map ~f:process_decl t
+  | _ -> failwith "Error: Unexpected form in external smt model"
+
+(* We are still missing some funcdecls, particularly function return values *)
+(** [check_external] invokes an external smt solver as a process. It communicates to the
+    process via a fairly rigid interpretation of the smtlib2 protocol. It extracts the
+    smtlib2 query string from the given z3_solver, queries the external solver, receives
+    a counter model if SAT. It then asserts this model back to the z3_solver, which should
+    check quickly. This solver can then be used with the regular cbat z3 model
+    interpretation machinery. Tested with Boolector, results with other solvers may vary.*)
+
+let check_external
+    (solver : Z3.Solver.solver)
+    (solver_path : string)
+    (ctx : Z3.context)
+    (declsyms : (Z3.FuncDecl.func_decl * Z3.Symbol.symbol) list) : Z3.Solver.status =
+  let smt_string = Z3.Solver.to_string solver in
+  let smt_preamble = "(set-logic QF_AUFBV) (set-option :produce-models true)" in
+  let smt_postamble = "(check-sat) (get-model) (exit)" in
+  (* Todo: Forward verbose flags to external solver? *)
+  let (solver_stdout, solver_stdin) = Caml_unix.open_process solver_path in
+  (* Send query to solver *)
+  Out_channel.output_string solver_stdin smt_preamble;
+  Out_channel.output_string solver_stdin smt_string;
+  Out_channel.output_string solver_stdin smt_postamble;
+  Out_channel.flush solver_stdin;
+  printf "Running external solver %s\n%!" solver_path;
+
+  (* SexpLib unfortunately uses # as an comment delimitter.
+     We replace it with a special token and revert this after parsing. *)
+  let pound_token = "MYSPECIALPOUND682" in
```

Comment:
Does this string have any special meaning? Or is it an arbitrary name only used for parsing?

Reply:
No it has no special meaning. This extremely embarrassing section is a hack to try and work around the incompatibility of ocaml Sexp and smtlib

Reply:
I can't think of a better way to handle this. I think the comment is sufficient enough for now.